app/src/androidTest/java/com/duckduckgo/app/fire/DataClearerTimeKeeperTest.kt

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2018 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.fire
+
+import com.duckduckgo.app.settings.clear.ClearWhenOption
+import com.duckduckgo.app.settings.clear.ClearWhenOption.*
+import com.duckduckgo.app.settings.db.SettingsDataStore
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.whenever
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import org.junit.runners.Parameterized.Parameters
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+@RunWith(Parameterized::class)
+class DataClearerTimeKeeperTest(private val testCase: TestCase) {
+
+    private val mockSettingsDataStore: SettingsDataStore = mock()
+    private val testee: DataClearerTimeKeeper = DataClearerTimeKeeper(mockSettingsDataStore)
+
+    companion object {
+
+        @JvmStatic
+        @Parameters(name = "Test case: {index} - {0}")
+        fun testData(): Array<TestCase> {
+
+            val timeNow = System.currentTimeMillis()
+
+            return arrayOf(
+                // no background time recorded - always expected to indicate not enough time passed
+                TestCase(false, APP_EXIT_ONLY, TimeUnit.MINUTES.toMillis(5), timeNow, false),
```

Comment:
The time **shouldn't** be relevant here, but the API requires the time. So these ensures that we get the expected result regardless of the time when it is `APP_EXIT_ONLY`; ie, the tests are validating that the time isn't relevant to the outcome.

Reply:
I might be missing something but when I look these two `APP_EXIT_ONLY` test cases:
- TestCase(false, APP_EXIT_ONLY, TimeUnit.MINUTES.toMillis(0), timeNow, false)
- TestCase(true, APP_EXIT_ONLY, TimeUnit.MINUTES.toMillis(5), timeNow)

It's not clear to me why they would behave differently depending on whether there is / isn't a background time recored, when the timing is irrelevant. I would expect them to both produce the same result.

Reply:
Yeah, I see what you're saying.

The check about whether the background time is recorded or not is more of a sanity check. So if we haven't yet recorded a timestamp for when the app is backgrounded, we bail early since the system is not in a known state. So those first few tests marked with `// no background time recorded` are really testing the guard is working.

You could argue that regardless of that, we'd know that for `APP_EXIT_ONLY` that the background time is irrelevant. But it feels stranger to allow that through despite us being in an unknown state; it feels safer in that case to bail out since a value we'd definitely expect to be set hasn't been.

Reply:
I think the second option is preferable out of the two. However, even if the logic of handling `APP_EXIT_ONLY` should be handled elsewhere, and the time keeper class shouldn't receive a `APP_EXIT_ONLY` option, the API permits it (since it's accepting any of the enums values).

Options for what to do if one is passed to the method:

1. throw an exception 
1. check it early and return `false` to bail early
1. don't explicitly check for it at all. in that case, the behaviour will be somewhat undefined.

I'd lean towards 2.

Reply:
I'm happy to go with 2.

Optional if you think it works and you have time: Accept the time as a param to `hasEnoughTimeElapsed` rather than operating on an enum which may or may not encode time information.

Reply:
Going with 2, we still need the `clearWhenOption: ClearWhenOption` option passed as a parameter, so i'd rather not also specify the time in addition to that.

Reply:
I've pushed some changes; let me know what you think / RR